rdm/0007-access-control.md

```diff
@@ -0,0 +1,222 @@
+    - Authenticated user: any logged in user.
+    - Any user: both authenticated and anonymous users.
+    - Campus user: Assigned to users based on their IP address being in a specific range.
+
+
+## Detailed design
+
+Invenio implements a Role-Based Access Control system (RBAC), that can be used to protect and grant access to resources.
+
+### Records as resources
+
+The most complex access rights use cases in Invenio RDM are related to records. Note that the following access control can be different between records and deposit records. In addition, this access control model is not tight to Invenio RDM, so it can be applied to any data model in Invenio.
+
+**Record actions**
+
+All records have the following basic CRUD actions:
+- Create: Ability to create new records.
+- Read: Ability to read a record (including during searches).
+- Update: Ability to update a record metadata.
+- Delete: Ability to delete a record.
+- List: Ability to perform searches (note, for a record to appear in the search result it is subject to the above mentioned read action).
+
+Records linked with files (possible for all records) have the additional following actions.
+- Create files: Ability to create files associated with a record. 
+- Read files: Ability to read files associated with a record.
+- Update files: Ability to update files associated with a record.
+- Delete files: Ability to delete files associated with a record.
+- Administrate files: Ability to perform all possible actions on files associated with a record, including reading previous versions of files.
+
+Deposit records have the additional following actions:
+- Publish: Ability to publish a new record.
+- Edit: Ability to edit an already published record. This differs from the record update action.
+- New version: Ability to create a new version of an existing record.
+
+It is possible to extend this model with more actions for each data model as needed.
+
+**Requiring permissions**
+
+A record must store permissions inside the metadata. This can be as simple as a list of owners, or very complex based on custom rules (e.g. books can only be curated by books curators). This rules are translated into `generators`.
+
+**Generators**
+
+A *generator* implements a rule to allow or deny the access to perform an action.  
+
+The generator would return a set of *needs* and *excludes* for each action. The *needs* are the constraints the user has to meet in order to be allowed to perform an action (e.g. be the owner of the record). The *excludes* specify a set of blacklisting constraints, which have precedence over the *needs*. 
+
+For example, the *record owners* generator would look like:
+
+``` python
+Document:
+    _owners: [1, 2, 3]
+
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+```
+
+Since the record owner does not set any specific blacklist (*exclude*), the generator would result in an access granted for users with id 1, 2 or 3.
```

Comment:
```suggestion
Since the record owner does not set any specific blacklist (*excludes*), the generator would result in an access granted for users with id 1, 2 or 3.
```